libraries/persistent_store/src/lib.rs

```diff
@@ -12,10 +12,336 @@
+//!     transaction use and free words as described above.
+//! -   `Clear` doesn't use capacity and frees the words used by the insertion of
+//!     the deleted entries.
+//! -   `Prepare` doesn't use capacity.
+//!
+//! The _total lifetime_ of the store is below `L = ((E + 1) * N - 1) * (P - 2)` and
+//! above `L - M` words, where `E` is the maximum number of erase cycles. The
+//! lifetime is used when capacity is used, including transiently, as well as when
+//! compaction occurs. The more the store is loaded (few remaining words of
+//! capacity), the more compactions are frequent, and the more lifetime is used.
+//!
+//! It is possible to approximate the cost of transient words in terms of capacity:
+//! `L` transient words are equivalent to `C - x` words of capacity where `x` is the
+//! average capacity (including transient) of operations.
+//!
+//! ## Preconditions
+//!
+//! The store may behave in unexpected ways if the following assumptions don't hold:
+//! -   A word can be written twice between erase cycles.
+//! -   A page can be erased `E` times after the first boot of the store.
+//! -   When power is lost while writing a slice or erasing a page, the next read
+//!     returns a slice where a subset (possibly none or all) of the bits that
+//!     should have been modified have been modified.
+//! -   Reading a slice is deterministic. When power is lost while writing a slice
+//!     or erasing a slice (erasing a page containing that slice), reading that
+//!     slice repeatedly returns the same result (until it is overwritten or its
+//!     page is erased).
+//! -   To decide whether a page has been erased, it is enough to test if all its
+//!     bits are equal to 1.
+//! -   When power is lost while writing a slice or erasing a page, that operation
+//!     does not count towards the limits. However, completing that write or erase
+//!     operation would count towards the limits, as if the number of writes per
+//!     word and number of erase cycles could be fractional.
+//! -   The storage is only modified by the store. Note that completely erasing the
+//!     storage is supported, essentially losing all content and lifetime tracking.
+//!     It is preferred to use `Clear` with a threshold of 0 to keep the lifetime
+//!     tracking.
+//!
+//! The store properties may still hold outside some of those assumptions but with
+//! weaker probabilities as the usage diverges from them.
+//!
+//! # Implementation
+//!
+//! We define the following constants:
+//! -   `E < 65536` the number of times a page can be erased.
+//! -   `3 <= N < 64` the number of pages in the storage.
+//! -   `8 <= P <= 1024` the number of words in a page.
+//! -   `Q = P - 2` the number of words in a virtual page.
+//! -   `K = 4096` the maximum number of keys.
+//! -   `M = min(Q - 1, 256)` the maximum length in words of a value.
+//! -   `V = (N - 1) * (Q - 1) - M` the virtual capacity.
+//! -   `C = V - N` the user capacity.
```

Comment:
Having unit tests checking `is_storage_supported` for practical instances sounds good to me. But some of these (in)equalities (or the fact that nothing overflows) don't obviously follow from the checks made by `is_storage_supported`, so it would be worth either (1) adding a proof that they indeed are ensured by `is_storage_supported` or (2) add them explicitly in `is_storage_supported` or (3) add unit tests for practical instances.

Reply:
I'm not sure which inequalities you talk about. The inequalities in this file are checked "as is" in `is_storage_supported`. I'm assuming you're talking about the "helper" inequalities in `Format`. Those are just given as lemmas to avoid having to show them each time for other more complicated proofs. Those lemmas only involve a few trivial steps on top the preceding lemmas (eventually ending up to the inequalities in this file which are those of `is_storage_supported`).

Fuzzing and unit tests make sure there are no under- and over-flows. The inequalities themselves are just useful for generic proofs like the one below.

Reply:
Resolving here to avoid multiplying threads, but the same remark applies: if the proof is easy, can you write down & document it?